File: src/main/java/io/myosotisdev/utopianism/util/Tags.java
package io.myosotisdev.utopianism.util;

import io.myosotisdev.utopianism.serialize.Serializer;
import net.minestom.server.tag.Tag;
import net.minestom.server.tag.TagReadable;

public class Tags
{
    public static Tag createTag(String key, Serializer serializer)
    {
        return Tag.Structure(key, serializer.createTagSerializer(key));
    }

    public static boolean compare(Tag key, TagReadable readable0, TagReadable readable1)
    {
        return readable0.getTag(key) == readable1.getTag(key);
    }

    public static <T> T getTagValue(Tag<T> tag, TagReadable readable)
    {
        return readable.getTag(tag);
    }
}

File: src/main/java/io/myosotisdev/utopianism/util/Tags.kt
package io.myosotisdev.utopianism.util

import io.myosotisdev.utopianism.serialize.Serializer
import net.minestom.server.tag.Tag
import net.minestom.server.tag.TagReadable

object Tags
{
    @JvmStatic
    fun <T> createTag(key: String, serializer: Serializer<T>): Tag<T>
    {
        return Tag.Structure(key, serializer.createTagSerializer(key))
    }

    fun compare(key: Tag<*>, readable0: TagReadable, readable1: TagReadable): Boolean
    {
        return readable0.getTag(key) === readable1.getTag(key)
    }

    fun <T> getTagValue(tag: Tag<T>, readable: TagReadable): T
    {
        return readable.getTag(tag)
    }
}